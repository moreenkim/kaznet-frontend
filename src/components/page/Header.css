.kaznet-root .dropdown-item.active,
.kaznet-root .dropdown-item:active {
  color: initial;
  background-color: inherit;
}

.dropdown-item .nav-link.active {
  border-bottom: none;
}
