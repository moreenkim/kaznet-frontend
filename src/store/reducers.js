import tasks from './tasks/reducer';

export {
  tasks
};