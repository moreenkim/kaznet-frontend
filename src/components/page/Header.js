// Renders the header
import React, { Component } from "react";
import { NavLink, Link, withRouter } from "react-router-dom";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import {
  Collapse,
  Navbar,
  NavbarToggler,
  Nav,
  NavItem,
  Container,
  Dropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem
} from "reactstrap";

import * as userSelectors from "../../store/users/reducer";
import * as userActions from "../../store/users/actions";

import profile_image from "../../images/profile.png";
import "./Header.css";

export class Header extends Component {
  componentDidMount() {
    this.props.fetchLoggedInUser();
  }

  constructor(props) {
    super(props);

    this.toggle = this.toggle.bind(this);
    this.state = {
      dropdownOpen: false
    };
  }

  toggle() {
    this.setState(prevState => ({
      dropdownOpen: !prevState.dropdownOpen
    }));
  }

  render() {
    const path = this.props.location.pathname;

    return (
      <header>
        <Container fluid>
          <Navbar light color="white" expand="md">
            <Link to="/" className="navbar-brand logo">
              KAZNET
            </Link>
            <NavbarToggler />
            <Collapse navbar>
              <Nav navbar>
                <NavItem>
                  <NavLink
                    to="/tasks"
                    className="nav-link"
                    activeClassName="active"
                  >
                    Tasks
                  </NavLink>
                </NavItem>
                <NavItem>
                  <NavLink
                    to="/clients"
                    className="nav-link"
                    activeClassName="active"
                  >
                    Clients
                  </NavLink>
                </NavItem>
                <NavItem>
                  <Dropdown
                    isOpen={this.state.dropdownOpen}
                    toggle={this.toggle}
                  >
                    <DropdownToggle
                      caret
                      tag="span"
                      className={
                        path === "/locations" || path === "/locationtypes"
                          ? "nav-link active"
                          : "nav-link"
                      }
                    >
                      Locations
                    </DropdownToggle>
                    <DropdownMenu>
                      <DropdownItem>
                        <NavLink
                          to="/locations"
                          className="nav-link"
                          activeClassName="active"
                        >
                          Locations
                        </NavLink>
                      </DropdownItem>
                      <DropdownItem>
                        <NavLink
                          to="/locationtypes"
                          className="nav-link"
                          activeClassName="active"
                        >
                          Location Types
                        </NavLink>
                      </DropdownItem>
                    </DropdownMenu>
                  </Dropdown>
                </NavItem>
                <NavItem>
                  <NavLink
                    to="/users"
                    className="nav-link"
                    activeClassName="active"
                  >
                    Users
                  </NavLink>
                </NavItem>
              </Nav>
              <Nav className="ml-auto" navbar>
                <NavItem>
                  <img
                    src={
                      this.props.getCurrentUser &&
                      this.props.getCurrentUser.attributes &&
                      this.props.getCurrentUser.attributes.metadata.gravatar
                        ? this.props.getCurrentUser.attributes.metadata.gravatar
                        : profile_image
                    }
                    className="img-fluid rounded-circle userprofile-img"
                    alt="profile"
                  />
                </NavItem>
              </Nav>
            </Collapse>
          </Navbar>
        </Container>
      </header>
    );
  }
}

function mapStateToProps(state, props) {
  return {
    getCurrentUser: userSelectors.getCurrentUser(state)
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      fetchLoggedInUser: userActions.fetchLoggedInUser
    },
    dispatch
  );
}

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(Header)
);
