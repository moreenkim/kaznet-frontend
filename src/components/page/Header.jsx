// Renders the header
import React, { Component } from 'react';
import { NavLink, Link, withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
  Collapse,
  Navbar,
  NavbarToggler,
  Nav,
  NavItem,
  Container,
  Dropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem
} from 'reactstrap';
import PropTypes from 'prop-types';

import * as userSelectors from '../../store/users/reducer';
import * as userActions from '../../store/users/actions';

import profileImage from '../../images/profile.png';
import * as constants from '../../constants';
import './Header.css';

export class Header extends Component {
  constructor(props) {
    super(props);

    this.toggleLocation = this.toggleLocation.bind(this);
    this.toggleTask = this.toggleTask.bind(this);
    this.toggleUser = this.toggleUser.bind(this);
    this.getAppName = this.getAppName.bind(this);
    this.state = {
      locationDropDownOpen: false,
      taskDropDownOpen: false,
      userDropDownOpen: false
    };
  }

  componentDidMount() {
    const { fetchLoggedInUser } = this.props;
    fetchLoggedInUser();
  }

  // eslint-disable-next-line class-methods-use-this
  getAppName() {
    if (constants.WEBSITE_NAME) return constants.WEBSITE_NAME.toUpperCase();
    return 'CERANA';
  }

  toggleLocation() {
    this.setState(prevState => ({
      locationDropDownOpen: !prevState.locationDropDownOpen
    }));
  }

  toggleTask() {
    this.setState(prevState => ({
      taskDropDownOpen: !prevState.taskDropDownOpen
    }));
  }

  toggleUser() {
    this.setState(prevState => ({
      userDropDownOpen: !prevState.userDropDownOpen
    }));
  }

  render() {
    const { location, getCurrentUser } = this.props;
    const path = location.pathname;

    const { taskDropDownOpen, locationDropDownOpen, userDropDownOpen } = this.state;

    return (
      <header>
        <Container fluid>
          <Navbar light color="white" expand="md">
            <Link to="/" className="navbar-brand logo">
              {this.getAppName()}
            </Link>
            <NavbarToggler />
            <Collapse navbar>
              <Nav navbar>
                <NavItem>
                  <Dropdown isOpen={taskDropDownOpen} toggle={this.toggleTask}>
                    <DropdownToggle
                      caret
                      tag="span"
                      className={
                        path === '/tasks' || path === '/forms' ? 'nav-link active' : 'nav-link'
                      }
                    >
                      Tasks
                    </DropdownToggle>
                    <DropdownMenu>
                      <DropdownItem>
                        <NavLink to="/tasks" className="nav-link" activeClassName="active">
                          Tasks
                        </NavLink>
                      </DropdownItem>
                      <DropdownItem>
                        <NavLink to="/forms" className="nav-link" activeClassName="active">
                          Forms
                        </NavLink>
                      </DropdownItem>
                    </DropdownMenu>
                  </Dropdown>
                </NavItem>
                <NavItem>
                  <NavLink to="/clients" className="nav-link" activeClassName="active">
                    Clients
                  </NavLink>
                </NavItem>
                <NavItem>
                  <Dropdown isOpen={locationDropDownOpen} toggle={this.toggleLocation}>
                    <DropdownToggle
                      caret
                      tag="span"
                      className={
                        path === '/locations' || path === '/locationtypes'
                          ? 'nav-link active'
                          : 'nav-link'
                      }
                    >
                      Locations
                    </DropdownToggle>
                    <DropdownMenu>
                      <DropdownItem>
                        <NavLink to="/locations" className="nav-link" activeClassName="active">
                          Locations
                        </NavLink>
                      </DropdownItem>
                      <DropdownItem>
                        <NavLink to="/locationtypes" className="nav-link" activeClassName="active">
                          Location Types
                        </NavLink>
                      </DropdownItem>
                    </DropdownMenu>
                  </Dropdown>
                </NavItem>
                <NavItem>
                  <NavLink to="/users" className="nav-link" activeClassName="active">
                    Users
                  </NavLink>
                </NavItem>
              </Nav>
              <Nav className="ml-auto" navbar>
                <NavItem>
                  <Dropdown isOpen={userDropDownOpen} toggle={this.toggleUser} size="sm">
                    <DropdownToggle tag="span" className="toggle-user">
                      <img
                        src={
                          getCurrentUser &&
                          getCurrentUser.attributes &&
                          getCurrentUser.attributes.metadata.gravatar
                            ? getCurrentUser.attributes.metadata.gravatar
                            : profileImage
                        }
                        className="img-fluid rounded-circle userprofile-img"
                        alt="profile"
                      />
                    </DropdownToggle>
                    {getCurrentUser && getCurrentUser.id && (
                      <DropdownMenu right>
                        <DropdownItem>
                          <NavLink
                            to={`/users/${getCurrentUser.id}`}
                            className="nav-link"
                            activeClassName="active"
                          >
                            View Profile
                          </NavLink>
                        </DropdownItem>
                        <DropdownItem>
                          <a href="/accounts/logout" className="nav-link">
                            Log Out
                          </a>
                        </DropdownItem>
                      </DropdownMenu>
                    )}
                  </Dropdown>
                </NavItem>
              </Nav>
            </Collapse>
          </Navbar>
        </Container>
      </header>
    );
  }
}

function mapStateToProps(state) {
  return {
    getCurrentUser: userSelectors.getCurrentUser(state)
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      fetchLoggedInUser: userActions.fetchLoggedInUser
    },
    dispatch
  );
}

Header.propTypes = {
  fetchLoggedInUser: PropTypes.func.isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string
  }).isRequired,
  getCurrentUser: PropTypes.shape({
    id: PropTypes.number,
    attributes: PropTypes.shape({
      metadata: PropTypes.shape({
        gravatar: PropTypes.string
      })
    })
  }).isRequired
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Header));
