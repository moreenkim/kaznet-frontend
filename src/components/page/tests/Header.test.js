// test Header
import React from "react";
import { shallow, mount } from "enzyme";
import { Router } from "react-router";
import createBrowserHistory from "history/createBrowserHistory";
import { createStore, applyMiddleware, combineReducers } from "redux";
import { connectRouter, routerMiddleware } from "connected-react-router";
import thunk from "redux-thunk";
import toJson from "enzyme-to-json";

import Header from "../Header";
import * as reducers from "../../../store/reducers";

const history = createBrowserHistory();

const store = createStore(
  connectRouter(history)(combineReducers(reducers)),
  applyMiddleware(thunk, routerMiddleware(history))
);

describe("components/page/Header", () => {
  it("renders without crashing", () => {
    shallow(<Header store={store} />);
  });

  it("renders header correctly", () => {
    const wrapper = mount(
      <Router history={history}>
        <Header store={store} />
      </Router>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
