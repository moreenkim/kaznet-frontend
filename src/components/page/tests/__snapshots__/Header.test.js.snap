// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/page/Header renders header correctly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Header
    profile_image="profile.png"
  >
    <header>
      <Container
        fluid={true}
        tag="div"
      >
        <div
          className="container-fluid"
        >
          <Navbar
            color="white"
            expand="md"
            light={true}
            tag="nav"
          >
            <nav
              className="navbar navbar-expand-md navbar-light bg-white"
            >
              <Link
                className="navbar-brand logo"
                replace={false}
                to="/"
              >
                <a
                  className="navbar-brand logo"
                  href="/"
                  onClick={[Function]}
                >
                  KAZNET
                </a>
              </Link>
              <NavbarToggler
                tag="button"
                type="button"
              >
                <button
                  className="navbar-toggler"
                  type="button"
                >
                  <span
                    className="navbar-toggler-icon"
                  />
                </button>
              </NavbarToggler>
              <Collapse
                appear={false}
                enter={true}
                exit={true}
                in={false}
                isOpen={false}
                mountOnEnter={false}
                navbar={true}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                tag="div"
                timeout={350}
                unmountOnExit={false}
              >
                <Transition
                  appear={false}
                  enter={true}
                  exit={true}
                  in={false}
                  mountOnEnter={false}
                  onEnter={[Function]}
                  onEntered={[Function]}
                  onEntering={[Function]}
                  onExit={[Function]}
                  onExited={[Function]}
                  onExiting={[Function]}
                  timeout={350}
                  unmountOnExit={false}
                >
                  <div
                    className="collapse navbar-collapse"
                    style={Object {}}
                  >
                    <Nav
                      navbar={true}
                      tag="ul"
                      vertical={false}
                    >
                      <ul
                        className="navbar-nav"
                      >
                        <NavItem
                          tag="li"
                        >
                          <li
                            className="nav-item"
                          >
                            <NavLink
                              activeClassName="active"
                              aria-current="page"
                              className="nav-link"
                              to="/tasks"
                            >
                              <Route
                                path="\\\\/tasks"
                              >
                                <Link
                                  aria-current={null}
                                  className="nav-link"
                                  replace={false}
                                  to="/tasks"
                                >
                                  <a
                                    aria-current={null}
                                    className="nav-link"
                                    href="/tasks"
                                    onClick={[Function]}
                                  >
                                    Tasks
                                  </a>
                                </Link>
                              </Route>
                            </NavLink>
                          </li>
                        </NavItem>
                        <NavItem
                          tag="li"
                        >
                          <li
                            className="nav-item"
                          >
                            <NavLink
                              activeClassName="active"
                              aria-current="page"
                              className="nav-link"
                              to="/clients"
                            >
                              <Route
                                path="\\\\/clients"
                              >
                                <Link
                                  aria-current={null}
                                  className="nav-link"
                                  replace={false}
                                  to="/clients"
                                >
                                  <a
                                    aria-current={null}
                                    className="nav-link"
                                    href="/clients"
                                    onClick={[Function]}
                                  >
                                    Clients
                                  </a>
                                </Link>
                              </Route>
                            </NavLink>
                          </li>
                        </NavItem>
                        <NavItem
                          tag="li"
                        >
                          <li
                            className="nav-item"
                          >
                            <NavLink
                              activeClassName="active"
                              aria-current="page"
                              className="nav-link"
                              to="/locations"
                            >
                              <Route
                                path="\\\\/locations"
                              >
                                <Link
                                  aria-current={null}
                                  className="nav-link"
                                  replace={false}
                                  to="/locations"
                                >
                                  <a
                                    aria-current={null}
                                    className="nav-link"
                                    href="/locations"
                                    onClick={[Function]}
                                  >
                                    Locations
                                  </a>
                                </Link>
                              </Route>
                            </NavLink>
                          </li>
                        </NavItem>
                        <NavItem
                          tag="li"
                        >
                          <li
                            className="nav-item"
                          >
                            <NavLink
                              activeClassName="active"
                              aria-current="page"
                              className="nav-link"
                              to="/users"
                            >
                              <Route
                                path="\\\\/users"
                              >
                                <Link
                                  aria-current={null}
                                  className="nav-link"
                                  replace={false}
                                  to="/users"
                                >
                                  <a
                                    aria-current={null}
                                    className="nav-link"
                                    href="/users"
                                    onClick={[Function]}
                                  >
                                    Users
                                  </a>
                                </Link>
                              </Route>
                            </NavLink>
                          </li>
                        </NavItem>
                      </ul>
                    </Nav>
                    <Nav
                      className="ml-auto"
                      navbar={true}
                      tag="ul"
                      vertical={false}
                    >
                      <ul
                        className="ml-auto navbar-nav"
                      >
                        <NavItem
                          tag="li"
                        >
                          <li
                            className="nav-item"
                          >
                            <img
                              alt="profile"
                              className="img-fluid rounded-circle userprofile-img"
                              src="profile.png"
                            />
                          </li>
                        </NavItem>
                      </ul>
                    </Nav>
                  </div>
                </Transition>
              </Collapse>
            </nav>
          </Navbar>
        </div>
      </Container>
    </header>
  </Header>
</Router>
`;
